Add support for TXT record type

arontsang opened the pull request:
**Description**

Adds support for deleting TXT records after the CRD is deleted.

<!-- Please link to all GitHub issue that this pull request implements(i.e. Fixes #123) -->
Fixes #5254 
Fixes #5279

**Checklist**

- [x] Unit tests updated
- [x] End user documentation updated

mloiseleur commented:
Thanks for this @arontsang :+1: 
Do you think you can add some tests on this ?

ivankatliarchuk commented:
Could you explain a little more what this code suppose to be doing? I do not understand it

arontsang commented:
Sure @ivankatliarchuk 

The bug.

https://github.com/kubernetes-sigs/external-dns/blob/master/registry/txt.go:198 is failing to apply the labels for TXT records to the TXT endpoints.

The result. 

https://github.com/kubernetes-sigs/external-dns/blob/master/controller/controller.go:255 `plan.Calculate()` regards any TXT records created by external-dns as not being created by external-dns (missing owner etc etc).

The cause.

a) Old style TXT records come in on the same record as the target

foo.acme.com TXT value [ 'FOOBAR', 'heritage=external-dns,....blah' ]

Depending on which value is returned first, we end up with two different but equally broken behaviors.

- We detect the record is an endpoint, but are missing the label
- We detect the label, but have no endpoint to apply it to....:(

Therefore we need to disable the creation of old style TXT records for TXT endpoints.
https://github.com/kubernetes-sigs/external-dns/blob/master/registry/txt.go:240

Also disable fallback to old style when annotating TXT endpoints.
https://github.com/kubernetes-sigs/external-dns/blob/master/registry/txt.go:199

b) New style TXT records are not recognized as being the heritage for the endpoint, because `getSupportedTypes()` is hard coded to NOT allow support for new style TXT records.
https://github.com/kubernetes-sigs/external-dns/blob/master/registry/txt.go:111

ivankatliarchuk commented:
I will try to give it another shot. But from my integration tests it does not work. The whole issue is, we support legacy and new formats, as a result the TXT handling is not properly supported.

Have you tried to do end-2-end testing and it works on your side?

ivankatliarchuk commented:
I've added comment in the issue https://github.com/kubernetes-sigs/external-dns/issues/5254#issuecomment-2800715310

arontsang commented:
I was debugging with a local cluster/docker with my own cloudflare account.

Seems to work well with deleting TXT records.

arontsang commented:
I also know that if the user has A/AAAA/CNAME records with the same name as the TXT records you are trying to create, my PR will not work.

It ONLY works with the case where TXT names are different from the other records, OR "only new TXT" is enabled.

Again, ideally, we inject the `--managed-record-types` values into `getSupportedTypes`.

I suggested both that, and a change where `--managed-record-types=TXT` requires that `--txt-new-format-only` is also switch on.

PS, did you test the change with  `--managed-record-types=TXT`?

ivankatliarchuk commented:
Indeed. It does not seems like CRD is care at all about this config. Seems like you have some other manifest that generates TXT record. As CRD simply does not support TXT from the code itself.

ivankatliarchuk commented:
> I suggested both that, and a change where --managed-record-types=TXT requires that --txt-new-format-only is also switch on.

This is it. The problem with TXT formats. There is a PR to simplify logic, https://github.com/kubernetes-sigs/external-dns/pull/5172 but most likely is not going into master anytime soon. CRD and TXT is a problem for long time.

Share all manifests. I could try once again. I'm testing on staging accounts where I have thousands of TXT recrods, not on fresh/empty

ivankatliarchuk commented:
> I suggested both that, and a change where --managed-record-types=TXT requires that --txt-new-format-only is also switch on.

This is it. The problem with TXT formats. There is a PR to simplify logic, https://github.com/kubernetes-sigs/external-dns/pull/5172 but most likely is not going into master anytime soon (in fact I'm planning to close it, as there are pushbucks and folks happy to have this TXT mess). CRD and TXT is a problem for long time.

Share all manifests. I could try once again. I'm testing on staging accounts where I have thousands of TXT recrods, not on fresh/empty

arontsang commented:
hey @ivankatliarchuk
I've updated my PR to fix the logic in the TXT record handling.

Previously it was assumed that a TXT records is either a heritage/label entry, OR a TXT entry.

We would only process the FIRST txt target for a TXT record into a label.

I've now changed the logic to process ALL targets as labels, and then process the rest as a TXT entry.

This means that we can use old style txt format with A/AAAA records together.
It should also mean that TXT should work correctly now.

mloiseleur commented:
/ok-to-test

arontsang commented:
/retest

mloiseleur commented:
@ivankatliarchuk anything left on your side ?
@mcharriere @mrozentsvayg Do you think one of you can proceed with a final review ?

ivankatliarchuk commented:
I could not approve. Both issues do not click to me. Overall I understand the direction, to have a better integration most likely with cert-manager or smth.

From the code perspective it's seems like a breaking change. We use to process first Target, now we process all record targets.

No integration tests before and after, unit tests not updated to cover a case.

ivankatliarchuk commented:
Let's keep it open. I'll asked one of the team working with to try on some close to real clusters. May takes few weeks
